In [1]:
!conda install -c conda-forge scikit-surprise

Solving environment: done

# All requested packages already installed.



In [3]:
from surprise import NormalPredictor, BaselineOnly, KNNBasic, KNNWithMeans,  KNNBaseline, SVD, SVDpp
from surprise import Dataset
from surprise.model_selection import cross_validate

# Load the movielens-100k dataset (download it if needed).
data = Dataset.load_builtin('ml-100k')

# Use the famous SVD algorithm.

norm_pred = NormalPredictor()
baseline_only = BaselineOnly()
knn_basic = KNNBasic()
knn_with_means = KNNWithMeans()
knn_baseline = KNNBaseline()
svd = SVD()
svd_pp = SVDpp()
# Run 5-fold cross-validation and print results.
cross_validate(norm_pred, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(baseline_only, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(knn_basic, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(knn_with_means, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(knn_baseline, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(svd, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)
cross_validate(svd_pp, data, measures=['RMSE', 'MAE', 'FCP'], cv=5, verbose=True)


Evaluating RMSE, MAE, FCP of algorithm NormalPredictor on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.5149  1.5149  1.5208  1.5208  1.5147  1.5172  0.0029  
MAE (testset)     1.2202  1.2140  1.2186  1.2223  1.2188  1.2188  0.0027  
FCP (testset)     0.5034  0.4975  0.4973  0.4986  0.4861  0.4966  0.0057  
Fit time          0.09    0.12    0.12    0.11    0.12    0.11    0.01    
Test time         0.10    0.10    0.10    0.10    0.19    0.12    0.03    
Estimating biases using als...
Estimating biases using als...
Estimating biases using als...
Estimating biases using als...
Estimating biases using als...
Evaluating RMSE, MAE, FCP of algorithm BaselineOnly on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9457  0.9513  0.9435  0.9427  0.9379  0.9442  0.0044  
MAE (testset)     0.7485  0.7531  0.7498  0.7489  0.7422  0.7485  0.0035  
FCP (testset)     0.7030  0.690

{&#39;test_rmse&#39;: array([0.91785677, 0.92808664, 0.92061762, 0.92556979, 0.91288353]),
 &#39;test_mae&#39;: array([0.72217461, 0.72494952, 0.72136172, 0.72863724, 0.7161946 ]),
 &#39;test_fcp&#39;: array([0.711361  , 0.70527626, 0.70542808, 0.71121048, 0.71359846]),
 &#39;fit_time&#39;: (139.5071828365326,
  142.66110515594482,
  144.75219988822937,
  141.42762994766235,
  135.42054295539856),
 &#39;test_time&#39;: (2.5089988708496094,
  2.793266773223877,
  2.7438762187957764,
  2.799530029296875,
  2.6305768489837646)}

---
Normal predictor
---

- RMSE (testset)    1.5149  1.5149  1.5208  1.5208  1.5147  1.5172  0.0029  
- MAE (testset)     1.2202  1.2140  1.2186  1.2223  1.2188  1.2188  0.0027  
- FCP (testset)     0.5034  0.4975  0.4973  0.4986  0.4861  0.4966  0.0057  
---
Baseline only
---

- RMSE (testset)    0.9457  0.9513  0.9435  0.9427  0.9379  0.9442  0.0044  
- MAE (testset)     0.7485  0.7531  0.7498  0.7489  0.7422  0.7485  0.0035  
- FCP (testset)     0.7030  0.6907  0.6995  0.6929  0.6969  0.6966  0.0044  
---
KNN Basic
---

- RMSE (testset)    0.9760  0.9785  0.9727  0.9804  0.9868  0.9789  0.0047  
- MAE (testset)     0.7689  0.7720  0.7700  0.7739  0.7801  0.7730  0.0040  
- FCP (testset)     0.7129  0.7066  0.7071  0.7164  0.7042  0.7094  0.0045  
---
KNN with Means
---

- RMSE (testset)    0.9502  0.9461  0.9596  0.9510  0.9410  0.9496  0.0061  
- MAE (testset)     0.7459  0.7455  0.7552  0.7524  0.7412  0.7480  0.0050  
- FCP (testset)     0.7015  0.7002  0.7063  0.7062  0.7049  0.7038  0.0025 
---
KNN Baseline
---

- RMSE (testset)    0.9272  0.9347  0.9327  0.9367  0.9211  0.9305  0.0057  
- MAE (testset)     0.7293  0.7355  0.7350  0.7381  0.7269  0.7329  0.0042  
- FCP (testset)     0.7080  0.7070  0.7039  0.7034  0.7177  0.7080  0.0052 
---
SVD
---

- RMSE (testset)    0.9408  0.9338  0.9317  0.9400  0.9310  0.9354  0.0041  
- MAE (testset)     0.7414  0.7362  0.7348  0.7411  0.7326  0.7372  0.0035  
- FCP (testset)     0.6978  0.6978  0.7033  0.6978  0.7028  0.6999  0.0026
---
SVDpp
---

- RMSE (testset)    0.9179  0.9281  0.9206  0.9256  0.9129  0.9210  0.0054  
- MAE (testset)     0.7222  0.7249  0.7214  0.7286  0.7162  0.7227  0.0041  
- FCP (testset)     0.7114  0.7053  0.7054  0.7112  0.7136  0.7094  0.0034    
---